File: modules/tests/src/commonTest/kotlin/arrays/ReturnArray.kt
@file:OptIn(ExperimentalUnsignedTypes::class)

package arrays

import natives.test.MyDictionary
import natives.test.MyEnum
import withLib
import kotlin.test.Test
import kotlin.test.assertContentEquals

class ReturnArray {

    @Test
    fun returnCharArray() = withLib {
        assertContentEquals(charArrayOf('a', 'b'), natives.test.returnCharArray())
    }

    @Test
    fun returnCharArrayN() = withLib {
        assertContentEquals(null, natives.test.returnCharArrayN())
    }

    @Test
    fun returnBooleanArray() = withLib {
        assertContentEquals(booleanArrayOf(true, false), natives.test.returnBooleanArray())
    }

    @Test
    fun returnByteArray() = withLib {
        assertContentEquals(byteArrayOf(1, 2), natives.test.returnByteArray())
    }

    @Test
    fun returnUByteArray() = withLib {
        assertContentEquals(ubyteArrayOf(1.toUByte(), UByte.MAX_VALUE), natives.test.returnUByteArray())
    }

    @Test
    fun returnShortArray() = withLib {
        assertContentEquals(shortArrayOf(1, 2), natives.test.returnShortArray())
    }

    @Test
    fun returnUShortArray() = withLib {
        assertContentEquals(ushortArrayOf(1.toUShort(), UShort.MAX_VALUE), natives.test.returnUShortArray())
    }

    @Test
    fun returnIntArray() = withLib {
        assertContentEquals(intArrayOf(1, 2), natives.test.returnIntArray())
    }

    @Test
    fun returnUIntArray() = withLib {
        assertContentEquals(uintArrayOf(1.toUInt(), UInt.MAX_VALUE), natives.test.returnUIntArray())
    }

    @Test
    fun returnLongArray() = withLib {
        assertContentEquals(longArrayOf(1, 2), natives.test.returnLongArray())
    }

    @Test
    fun returnULongArray() = withLib {
        assertContentEquals(ulongArrayOf(1.toULong(), ULong.MAX_VALUE), natives.test.returnULongArray())
    }

    @Test
    fun returnFloatArray() = withLib {
        assertContentEquals(floatArrayOf(1.1f, 2.2f), natives.test.returnFloatArray())
    }

    @Test
    fun returnDoubleArray() = withLib {
        assertContentEquals(doubleArrayOf(1.1, 2.2), natives.test.returnDoubleArray())
    }

    @Test
    fun returnStringArray() = withLib {
        assertContentEquals(arrayOf("string1", "string2"), natives.test.returnStringArray())
    }

    @Test
    fun returnStringArrayN() = withLib {
        assertContentEquals(arrayOf(null, null), natives.test.returnStringArrayN())
    }

    @Test
    fun returnEnumArray() = withLib {
        assertContentEquals(arrayOf(MyEnum.CASE1, MyEnum.CASE2), natives.test.returnEnumArray())
    }

    @Test
    fun returnDictionaryArray() = withLib {
        assertContentEquals(arrayOf(
            MyDictionary(1, 2, 3, 4),
            MyDictionary(5, 6, 7, 8)
        ), natives.test.returnDictionaryArray())
    }

    @Test
    fun returnDictionaryArrayN() = withLib {
        assertContentEquals(arrayOf(null, null), natives.test.returnDictionaryArrayN())
    }
}

File: modules/tests/natives/testcpp/src/arrays/return_array.cpp
#include <api.hpp>

KArray<uint16_t> return_char_array() {
    return KArray<uint16_t> {
        'a', 'b'
    };
}

KOptional<KArray<uint16_t>> return_char_array_n() {
    return KOptional<KArray<uint16_t>>();
}

KArray<bool> return_boolean_array() {
    return KArray {
        true, false
    };
}

KArray<int8_t> return_byte_array() {
    return KArray<int8_t> {
        1, 2
    };
}

KArray<uint8_t> return_ubyte_array() {
    return KArray<uint8_t> {
        1u, 255u
    };
}

KArray<int16_t> return_short_array() {
    return KArray<int16_t> {
        1, 2
    };
}

KArray<uint16_t> return_ushort_array() {
    return KArray<uint16_t> {
        1u, 65535u
    };
}

KArray<int32_t> return_int_array() {
    return KArray {
        1, 2
    };
}

KArray<uint32_t> return_uint_array() {
    return KArray {
        1u, 4294967295u
    };
}

KArray<int64_t> return_long_array() {
    return KArray<int64_t> {
        1, 2
    };
}

KArray<uint64_t> return_ulong_array() {
    return KArray<uint64_t> {
        1u, 18446744073709551615u
    };
}

KArray<float> return_float_array() {
    return KArray {
        1.1f, 2.2f
    };
}

KArray<double> return_double_array() {
    return KArray {
        1.1, 2.2
    };
}

KArray<KString> return_string_array() {
    return KArray {
        KString("string1"), KString("string2")
    };
}

KArray<KOptional<KString>> return_string_array_n() {
    return KArray {
        KOptional<KString>(),
        KOptional<KString>()
    };
}

KArray<MyEnum> return_enum_array() {
    return KArray {
        CASE1, CASE2
    };
}

KArray<MyDictionary> return_dictionary_array() {
    return KArray {
        MyDictionary(1, 2, 3, 4),
        MyDictionary(5, 6, 7, 8)
    };
}

KArray<KOptional<MyDictionary>> return_dictionary_array_n() {
    return KArray {
        KOptional<MyDictionary>(),
        KOptional<MyDictionary>()
    };
}